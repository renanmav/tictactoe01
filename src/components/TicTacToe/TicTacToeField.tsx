import React from 'react';
import { Image, TouchableWithoutFeedback, View } from 'react-native';
import { t } from 'react-native-tailwindcss';

const s = {
  wrapper: [t.bgTransparent, t.w4_12, t.p1, { aspectRatio: 1 }],
  field: [t.bgBlue900, t.rounded, t.flex1, t.itemsCenter, t.justifyCenter, t.overflowHidden],
  symbol: [t.w8_12, t.objectContain],
  success: [t.bgGreen900],
};

interface TicTacToeFieldProps {
  symbol: 'X' | 'O' | ' ';
  onPress: () => void;
  success?: boolean;
}

const TicTacToeField: React.FC<TicTacToeFieldProps> = ({ symbol, onPress, success = false }) => {
  function renderSymbol() {
    switch (symbol) {
      case 'X':
        return <Image source={require('../../../assets/x.png')} style={s.symbol} testID="X" />;
      case 'O':
        return <Image source={require('../../../assets/o.png')} style={s.symbol} testID="O" />;
      default:
        return null;
    }
  }

  return (
    <View style={s.wrapper}>
      <TouchableWithoutFeedback onPress={onPress} testID="field-touchable">
        <View style={[s.field, success && s.success]} testID="field-background">
          {renderSymbol()}
        </View>
      </TouchableWithoutFeedback>
    </View>
  );
};

export default TicTacToeField;
