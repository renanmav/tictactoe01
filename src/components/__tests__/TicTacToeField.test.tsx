import React from 'react';
import { fireEvent, render } from '@testing-library/react-native';
import { t } from 'react-native-tailwindcss';

import TicTacToeField from '../TicTacToe/TicTacToeField';

it('should render the correct symbol', async () => {
  const X = render(<TicTacToeField symbol="X" onPress={jest.fn} />);
  const O = render(<TicTacToeField symbol="O" onPress={jest.fn} />);
  const Empty = render(<TicTacToeField symbol=" " onPress={jest.fn} />);

  expect(X.getByTestId('X')).toMatchSnapshot();
  expect(O.getByTestId('O')).toMatchSnapshot();
  expect(Empty.getByTestId('field-background')).toMatchSnapshot();
});

it('should render with different background if success prop is passed', async () => {
  const { getByTestId } = render(<TicTacToeField symbol="X" onPress={jest.fn} success />);

  expect(getByTestId('field-background')).toHaveStyle(t.bgGreen900);
});

it('should call onPress when pressed', async () => {
  const onPress = jest.fn();

  const { getByTestId } = render(<TicTacToeField symbol="X" onPress={onPress} success />);
  fireEvent.press(getByTestId('field-touchable'));

  expect(onPress).toBeCalledTimes(1);
});
