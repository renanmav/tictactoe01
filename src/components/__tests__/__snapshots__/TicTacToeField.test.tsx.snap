// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the correct symbol 1`] = `
<Image
  source={1}
  style={
    Array [
      Object {
        "width": "66.666667%",
      },
      Object {
        "resizeMode": "contain",
      },
    ]
  }
  testID="X"
/>
`;

exports[`should render the correct symbol 2`] = `
<Image
  source={1}
  style={
    Array [
      Object {
        "width": "66.666667%",
      },
      Object {
        "resizeMode": "contain",
      },
    ]
  }
  testID="O"
/>
`;

exports[`should render the correct symbol 3`] = `
<View
  style={
    Array [
      Array [
        Object {
          "backgroundColor": "#242c3a",
        },
        Object {
          "borderRadius": 4,
        },
        Object {
          "flex": 1,
        },
        Object {
          "alignItems": "center",
        },
        Object {
          "justifyContent": "center",
        },
        Object {
          "overflow": "hidden",
        },
      ],
      false,
    ]
  }
  testID="field-background"
/>
`;
